services/motion/builtin/state/state.go

```diff
@@ -52,16 +67,529 @@ type PlannerExecutor interface {
+				}
+
+				e.notifyStateRePlan(lastPWE.plan, res.resp.ReplanReason, newPWE.plan, time.Now())
+				lastPWE = newPWE
+			}
+		}
+	})
+
+	return nil
+}
+
+func (e *execution[R]) toStateExecution() stateExecution {
+	return stateExecution{
+		id:            e.id,
+		componentName: e.componentName,
+		waitGroup:     e.waitGroup,
+		cancelFunc:    e.cancelFunc,
+	}
+}
+
+func (e *execution[R]) notifyStateNewExecution(execution stateExecution, plan motion.Plan, time time.Time) {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	// NOTE: We hold the lock for both updateStateNewExecution & updateStateNewPlan to ensure no readers
+	// are able to see a state where the execution exists but does not have a plan with a status.
+	e.state.updateStateNewExecution(execution)
+	e.state.updateStateNewPlan(newPlanMsg{
+		plan:       plan,
+		planStatus: motion.PlanStatus{State: motion.PlanStateInProgress, Timestamp: time},
+	})
+}
+
+func (e *execution[R]) notifyStateRePlan(lastPlan motion.Plan, reason string, newPlan motion.Plan, time time.Time) {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	// NOTE: We hold the lock for both updateStateNewExecution & updateStateNewPlan to ensure no readers
+	// are able to see a state where the old plan is failed withou a new plan in progress during replanning
+	e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        lastPlan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateFailed, Timestamp: time, Reason: &reason},
+	})
+
+	e.state.updateStateNewPlan(newPlanMsg{
+		plan:       newPlan,
+		planStatus: motion.PlanStatus{State: motion.PlanStateInProgress, Timestamp: time},
+	})
+}
+
+func (e *execution[R]) notifyStatePlanFailed(plan motion.Plan, reason string, time time.Time) {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateFailed, Timestamp: time, Reason: &reason},
+	})
+}
+
+func (e *execution[R]) notifyStatePlanSucceeded(plan motion.Plan, time time.Time) {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateSucceeded, Timestamp: time},
+	})
+}
+
+func (e *execution[R]) notifyStatePlanStopped(plan motion.Plan, time time.Time) {
+	e.state.mu.Lock()
+	defer e.state.mu.Unlock()
+	e.state.updateStateStatusUpdate(stateUpdateMsg{
+		componentName: e.componentName,
+		executionID:   e.id,
+		planID:        plan.ID,
+		planStatus:    motion.PlanStatus{State: motion.PlanStateStopped, Timestamp: time},
+	})
+}
+
 // State is the state of the builtin motion service
 // It keeps track of the builtin motion service's executions.
-type State struct{}
+type State struct {
```

Comment:
This feels like it should be private, is there a reason its public?

Reply:
B/c the go linter complains if we return a private type from a `NewState` All its members are private. Its basically just an optic handle